File: backend/src/main/java/com/scm/service/interfaces/ContactService.java
package com.scm.service.interfaces;

import com.scm.entity.Contact;
import org.springframework.data.domain.Page;
import org.springframework.web.multipart.MultipartFile;

public interface ContactService {
    Contact createContact(String userId, Contact contact, MultipartFile image) throws Exception;
    Page<Contact> getContacts(String userId, int page, int size);
    Page<Contact> getFavoriteContacts(String userId, int page, int size);
    Contact getContactById(String userId, String contactId);
    Contact updateContact(String userId, String contactId, Contact updated, MultipartFile image) throws Exception;
    void deleteContact(String userId, String contactId);
    Page<Contact> searchContacts(String userId, String field, String keyword, int page, int size);
    long getContactCount(String userId);
    long getFavoriteContactCount(String userId);
    java.util.List<java.util.Map<String, Object>> getGroupStats(String userId);
    java.util.List<java.util.Map<String, Object>> getWeeklyStats(String userId);
}

File: frontend/src/services/contactService.js
import api from './api'

const buildFormData = (contactData, imageFile) => {
  const fd = new FormData()
  fd.append('contact', new Blob([JSON.stringify(contactData)], { type: 'application/json' }))
  if (imageFile) fd.append('image', imageFile)
  return fd
}

const contactService = {
  getContacts: (page = 0, size = 9) =>
    api.get('/contacts', { params: { page, size } }).then(r => r.data),

  getFavorites: (page = 0, size = 9) =>
    api.get('/contacts/favorites', { params: { page, size } }).then(r => r.data),

  getContact: (id) =>
    api.get(`/contacts/${id}`).then(r => r.data),

  createContact: (contactData, imageFile) =>
    api.post('/contacts', buildFormData(contactData, imageFile), {
      headers: { 'Content-Type': 'multipart/form-data' },
    }).then(r => r.data),

  updateContact: (id, contactData, imageFile) =>
    api.put(`/contacts/${id}`, buildFormData(contactData, imageFile), {
      headers: { 'Content-Type': 'multipart/form-data' },
    }).then(r => r.data),

  deleteContact: (id) =>
    api.delete(`/contacts/${id}`).then(r => r.data),

  searchContacts: (field, keyword, page = 0, size = 9) =>
    api.get('/contacts/search', { params: { field, keyword, page, size } }).then(r => r.data),

  getCount: () =>
    api.get('/contacts/count').then(r => r.data),

  getFavoritesCount: () =>
    api.get('/contacts/favorites/count').then(r => r.data),

  getGroupStats: () =>
    api.get('/contacts/group-stats').then(r => r.data),

  getWeeklyStats: () =>
    api.get('/contacts/weekly-stats').then(r => r.data),
}

export default contactService
